exchanges/subscription/list_test.go

```diff
@@ -45,3 +67,154 @@ func TestListGroupPairs(t *testing.T) {
+	assert.Equal(t, SubscribedState, l[1].State(), "SetState should set State correctly after the error")
+}
+
+// TestExpandTemplates exercises ExpandTemplates
+func TestExpandTemplates(t *testing.T) {
+	t.Parallel()
+	l := &List{
+		{Channel: CandlesChannel,
+			Template: "ohlc.{{$asset}}.{{$s.Interval.Short}}",
+			Asset:    asset.All,
+			Pairs:    currency.Pairs{btcusdtPair, ethusdcPair},
+			Interval: kline.FifteenMin},
+		{Channel: OrderbookChannel,
+			Template: "book-{{$pair}}@{{$s.Levels}}",
+			Asset:    asset.Spot,
+			Pairs:    currency.Pairs{btcusdtPair, ethusdcPair},
+			Levels:   100},
+		{Channel: CandlesChannel,
+			Template: "candles.{{assetName $asset}}.{{$pair.Swap.String}}.{{if eq $pair.String `BTCUSDT`}}{{$s.Params.color}}{{else}}red{{end}}.{{$s.Interval.Short}}",
+			Asset:    asset.All,
+			Pairs:    currency.Pairs{btcusdtPair, ethusdcPair},
+			Params: map[string]any{
+				"color": "green",
+			},
+			Interval: kline.FifteenMin},
+		{Channel: MyTradesChannel,
+			Template:      "trades.{{assetName $asset}}",
+			Asset:         asset.All,
+			Pairs:         currency.Pairs{btcusdtPair, ethusdcPair},
+			Authenticated: true,
+		},
+	}
+	got, err := l.ExpandTemplates(&mockEx{false})
+	require.NoError(t, err, "ExpandTemplates must not error")
+	exp := List{
+		{Channel: CandlesChannel, QualifiedChannel: "ohlc.spot.15m", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}, Interval: kline.FifteenMin},
+		{Channel: CandlesChannel, QualifiedChannel: "ohlc.futures.15m", Asset: asset.Futures, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}, Interval: kline.FifteenMin},
+		{Channel: OrderbookChannel, QualifiedChannel: "book-BTCUSDT@100", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair}, Levels: 100},
+		{Channel: OrderbookChannel, QualifiedChannel: "book-ETHUSDC@100", Asset: asset.Spot, Pairs: currency.Pairs{ethusdcPair}, Levels: 100},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.spot.USDTBTC.green.15m", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.spot.USDCETH.red.15m", Asset: asset.Spot, Pairs: currency.Pairs{ethusdcPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.future.USDTBTC.green.15m", Asset: asset.Futures, Pairs: currency.Pairs{btcusdtPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+		{Channel: CandlesChannel, QualifiedChannel: "candles.future.USDCETH.red.15m", Asset: asset.Futures, Pairs: currency.Pairs{ethusdcPair}, Interval: kline.FifteenMin,
+			Params: map[string]any{"color": "green"}},
+	}
+
+	if !equalLists(t, exp, got) {
+		t.FailNow() // If the first list isn't equal testing it again will duplicate test failures
+	}
+
+	got, err = l.ExpandTemplates(&mockEx{true})
```

Comment:
Just a thing that we might consider in the future making Expand templates agnostic to auth and just spew out everything, It's probably a pain so just verbose thoughts.

Reply:
I thought about this the same, that generateSubs should/could be agnostic to auth or not, and it's the actual Subscribe that should care.
But then I realised that if you look at it as a separation of powers, and generate's job is to fan out the config into the concretes based on the runtime context, then we're already doing the right thing.

The concept I'm trying to avoid (now) is that we have Features.Subscriptions that don't match what we'll (probably) actually AddSubscription for, so that diff works nicely, and just for sanity.
In time I want to have:
* Config is what's configured, nothing else
* Config Defaults are available by unzipping the config context (more on this later)
```
```
* Features.Subs contains what we expect to run
* Websocket.Subs is what's been manifested from that

So yeah, as it stands, if we change pairs, assets or auth, then we re-call generate and go again.

One gotcha I'm struggling with is that manual calls to Subscribe start to get really messy, because they bypass fan-out operations. I'm tempted that we privatise the flow so that uses *must* go through a Set Features->generate->Subscribe process which is automatic. I'm going to come back to this when the dust settles on config generally though.

Reply:
Privatising the flow would be much better. Thanks for the explanation, looks good.